### PR TITLE
JBPM-5743: Add css layout customization for Forms Fields

### PR DESCRIPTION
**Pull request by jesuino**
Implementation for https://issues.jboss.org/browse/JBPM-5743

Part of an ensemble:

kie-wb-common: https://github.com/kiegroup/kie-wb-common/pull/2618

### CODE REVIEW

**Comment by jesuino** [comment]
cc @paulovmr would you please review?

Hello @barboras7  - here's the PR for BAPL-1159 - more information to come soon (video and blog post) - would you please review when you have some time?

**Comment by jesuino** [comment]
Jenkins execute full downstream build

**Comment by pefernan** [comment]
@jesuino I'll start reviewing tomorrow, but there might be something wrong on the migration tool, looks like the tests are failing. Tomorrow i'll dig a little and I'll give you some info.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java [hunk 1]
```diff
@@ -26,7 +26,7 @@
 @Portable
 public final class LayoutTemplate {
 
-    private int version = 2;
+    private int version = 3;
```
Comment (pefernan) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java]: We need that on the form definition aswell!

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java [hunk 2]
```diff
@@ -414,6 +431,11 @@ public void onSelected() {
             }
         }
     }
+    
+    @Override
+    public List<LayoutEditorElementPart> getLayoutEditorElementParts() {
+        return parts;
+    }
 
     public interface View extends UberElement<ComponentColumn> {
```
Comment (pefernan) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java]: I'd add a @PreDestroy to clear the LayoutColumnParts generated with the ManagedInstance

**Comment by pefernan** [comment]
downstream build is failing due to some errors on other modules, retrigering it!

**Comment by pefernan** [comment]
jenkins execute full downstream build please

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponent.java [hunk 3]
```diff
@@ -39,52 +48,101 @@ public String getDragTypeName() {
...
+    
+    public List<LayoutComponentPart> getParts() {
+        return Collections.unmodifiableList(parts);
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
+        result = prime * result + ((dragTypeName == null) ? 0 : dragTypeName.hashCode());
```
Comment (paulovmr) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponent.java]: If this method is called in client-side, you have to do `result = ~~result` between each result attribution due to a GWT limitation.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponentPart.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    
+    public void removeCssProperty(String property) {
+        cssProperties.remove(property);
+        
+    }
+
+    @Override
+    public int hashCode() {
+        final int prime = 31;
+        int result = 1;
```
Comment (paulovmr) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponentPart.java]: Same here.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java (hunk 2 above)
Comment (paulovmr) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java]: @pefernan @jesuino `ManagedInstance` keeps track of the create instances and destroy them automatically when the bean instance in which it is injected is destroyed, so no need for `@PreDestroy` here.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/widgets/LayoutEditorPropertiesPresenter.java [hunk 5]
```diff
@@ -296,4 +314,39 @@ void onComponentRemoved(@Observes ComponentRemovedEvent event) {
...
+            currentLayoutElementParts.stream().map(LayoutEditorElementPart::getId)
+                                             .forEach(parts::add);
+            view.showParts(parts);
+        }
+    }
+    
+    private void updateCurrentPropertiesPresenter(String elementId, LayoutElementWithProperties layoutElement) {
+        LayoutElementPropertiesPresenter presenter = presenterMap.get(elementId);
+        if (presenter == null) {
+            propertiesPresenter = beanManager.lookupBean(LayoutElementPropertiesPresenter.class).newInstance();
```
Comment (paulovmr) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/widgets/LayoutEditorPropertiesPresenter.java]: Could you change this to use `ManagedInstance` instead?

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java (hunk 1 above)
Comment (jesuino) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java]: Hello @pefernan  - you mean we should add this field to FormDefinition and start counting from 1?

**Comment by jesuino** [comment]
@paulovmr @pefernan The changes are in latest commit - thanks!

**Comment by jesuino** [comment]
Jenkins execute full downstream build

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java (hunk 1 above)
Comment (pefernan) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutTemplate.java]: @jesuino sorry, this comment isn't a change suggestion just a comment :D

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java (hunk 2 above)
Comment (pefernan) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/main/java/org/uberfire/ext/layout/editor/client/components/columns/ComponentColumn.java]: @paulovmr good to know that, I always clear the ManagedInstances with  @Predestroys... so @jesuino forget this comment

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponent.java (hunk 3 above)
Comment (pefernan) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponent.java]: +1

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponentPart.java (hunk 4 above)
Comment (pefernan) [reply uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-api/src/main/java/org/uberfire/ext/layout/editor/api/editor/LayoutComponentPart.java]: +1

**Comment by barboras7** [comment]
Hi @jesuino,
I found one issue with the newly added panel (see [video](https://drive.google.com/file/d/1Yovs_Sny95fq7bAfy3w9lW263J4bZf5V/view?usp=sharing)). However, I don't think that this issue should prevent this PR from merging and if we agree on that, I can just create a JIRA for that (and I believe the issue will be present for other kinds of assets and their panels as well). 

As you can see in the video, if you create a form, then open some other asset and open the form again, the panel is not visible anymore and there is no button to expand this panel.

**Comment by jesuino** [comment]
Hello @barboras7  - thanks for letting me know about this issue, I think Pere mentioned it before as well. I agree to merge this and I can work on the dock pane problem on a separated JIRA.

**Comment by barboras7** [comment]
Thanks, @jesuino! I created a Jira and assigned it to you, see [AF-1962](https://issues.jboss.org/browse/AF-1962)

**Comment by jesuino** [comment]
for this @paulovmr looks like there's an unrelated failing test -> https://rhba-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/KIE/job/master/job/pullrequest/job/appformer-downstream-pullrequests/260/

**Pull request merged**